beacon-chain/forkchoice/protoarray/node.go

```diff
@@ -34,17 +48,128 @@ func (n *Node) Weight() uint64 {
 	return n.weight
 }
 
-// BestChild of the fork choice node.
-func (n *Node) BestChild() uint64 {
-	return n.bestChild
+// Children returns the children of this node
+func (n *Node) Children() []*Node {
+	return n.children
+}
+
+// depth returns the length of the path to the root of Fork Choice
+func (n *Node) depth() uint64 {
+	ret := uint64(0)
+	for node := n.parent; node != nil; node = node.parent {
+		ret += 1
+	}
+	return ret
+}
+
+// applyWeightChanges recomputes the weight of the node passed as an argument and all of its descendants,
+// using the current balance stored in each node. This function requires a lock
+// in Store.nodesLock
+func (n *Node) applyWeightChanges(ctx context.Context) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+	// update the children
+	childrensWeight := uint64(0)
+	for _, child := range n.children {
+		if err := child.applyWeightChanges(ctx); err != nil {
+			return err
+		}
+		childrensWeight += child.weight
+	}
+	if n.root == params.BeaconConfig().ZeroHash {
+		return nil
+	}
```

Comment:
I think we can move this to the start of the function but some tests failed so I didn't change it msyelf

Reply:
no, because for example at genesis you will have the root element being precisely that, and you will be stopping early without updating the children before. I guess that's what you see in those tests